adapter/internal/discovery/xds/common/utils.go

```diff
@@ -0,0 +1,96 @@
+/*
+ *  Copyright (c) 2022, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+// Package common includes the common functions shared between enforcer and router callbacks.
+package common
+
+import (
+	"sync"
+
+	discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+)
+
+const nodeIDArrayMaxLength int = 20
+const instanceIdentifierKey string = "instanceIdentifier"
+
+// NodeQueue struct is used to keep track of the nodes connected via the XDS.
+type NodeQueue struct {
+	lock  *sync.Mutex
+	queue []string
+}
+
+// CheckEntryAndSwapToEnd function does the following. Recently accessed entry is removed last.
+// Array should have a maximum length. If the the provided nodeId may or may not be within the array.
+//
+// 1. If the array's maximum length is not reached after adding the new element and the element is not inside the array,
+// 		append the element to the end.
+// 2. If the array is at maximum length and element is not within the array, the new entry should be appended to the end
+//		and the 0th element should be removed.
+// 3. If the array is at the maximum length and element is inside the array, the new element should be appended and the already
+// 		existing entry should be removed from the position.
+// Returns the modified array and true if the entry is a new addition.
+func CheckEntryAndSwapToEnd(array []string, nodeID string) (modifiedArray []string, isNewAddition bool) {
+	matchedIndex := -1
+	arraySize := len(array)
+	for index := arraySize - 1; index >= 0; index-- {
+		entry := array[index]
+		if entry == nodeID {
+			matchedIndex = index
+			break
+		}
+	}
+
+	if matchedIndex == nodeIDArrayMaxLength-1 {
+		return array, false
+	} else if matchedIndex > 0 {
+		array = append(array[0:matchedIndex], array[matchedIndex+1:]...)
+		array = append(array, nodeID)
+		return array, false
+	}
+	if arraySize >= nodeIDArrayMaxLength {
+		array = array[1:]
+	}
+	array = append(array, nodeID)
+	return array, true
+}
+
+// GenerateNodeQueue creates an instance of nodeQueue with a mutex and a string array assigned.
+func GenerateNodeQueue() *NodeQueue {
+	return &NodeQueue{
+		lock:  &sync.Mutex{},
+		queue: []string{},
+	}
+}
+
+// IsNewNode returns true if the provided nodeID does not exist in the nodeQueue
+func IsNewNode(nodeQueueInstance *NodeQueue, nodeIdentifier string) bool {
```

Comment:
Shall we associate this function with the struct NodeQueue? so it is a method of the struct

Reply:
fixed using ` func (nodeQueue *NodeQueue) IsNewNode(nodeID string) bool {..} `  via https://github.com/wso2/product-microgateway/pull/2876/commits/819883bef0563bbcb4135ec42df395fdb20c92a7